pulsar-broker/src/main/java/org/apache/pulsar/broker/service/SystemTopicBaseTxnBufferSnapshotService.java

```diff
@@ -18,69 +18,163 @@
+        AtomicBoolean exitingFlag = new AtomicBoolean(false);
+        ReferenceCountedWriter referenceCountedWriter = writerMap.compute(namespaceName, (k, v) -> {
+            if (v == null) {
+                return new ReferenceCountedWriter(namespaceName,
+                        getTransactionBufferSystemTopicClient(namespaceName).newWriterAsync(), writerMap);
+            }
+            exitingFlag.set(true);
+            return v;
+        });
+        if (exitingFlag.get()) {
+            referenceCountedWriter.retain();
+        }
+        return referenceCountedWriter;
+    }
+
+    @Override
+    public synchronized void releaseReferenceWriter(ReferenceCountedWriter referenceCountedWriter) {
+        if (referenceCountedWriter.release() == 0) {
+            writerMap.remove(referenceCountedWriter.namespaceName, referenceCountedWriter);
+            referenceCountedWriter.future.thenAccept(writer -> {
+                pendingCloseWriterList.add(writer);
+                closePendingCloseWriter();
+            });
         }
-        return CompletableFuture.completedFuture(clients.computeIfAbsent(systemTopicName,
+    }
+
+    private SystemTopicClient<TransactionBufferSnapshot> getTransactionBufferSystemTopicClient(
+            NamespaceName namespaceName) {
+        return clients.computeIfAbsent(namespaceName,
                 (v) -> namespaceEventsSystemTopicFactory
-                        .createTransactionBufferSystemTopicClient(topicName.getNamespaceObject(), this)));
+                        .createTransactionBufferSystemTopicClient(namespaceName, this));
     }
 
     @Override
     public CompletableFuture<Reader<TransactionBufferSnapshot>> createReader(TopicName topicName) {
-        return getTransactionBufferSystemTopicClient(topicName).thenCompose(SystemTopicClient::newReaderAsync);
+        return getTransactionBufferSystemTopicClient(topicName.getNamespaceObject()).newReaderAsync();
     }
 
     @Override
     public void removeClient(TopicName topicName,
                                           TransactionBufferSystemTopicClient transactionBufferSystemTopicClient) {
         if (transactionBufferSystemTopicClient.getReaders().size() == 0
                 && transactionBufferSystemTopicClient.getWriters().size() == 0) {
-            clients.remove(topicName);
+            clients.remove(topicName.getNamespaceObject());
         }
     }
 
     @Override
     public void close() throws Exception {
-        for (Map.Entry<TopicName, SystemTopicClient<TransactionBufferSnapshot>> entry : clients.entrySet()) {
+        for (Map.Entry<NamespaceName, SystemTopicClient<TransactionBufferSnapshot>> entry : clients.entrySet()) {
             entry.getValue().close();
         }
     }
+
+    private void closePendingCloseWriter() {
+        Iterator<Writer<TransactionBufferSnapshot>> iterator = pendingCloseWriterList.stream().iterator();
```

Comment:
Nit: why are you creating the stream?
We can iterate on the LinkedList directly, creating less garbage  and also making the code simpler